content/posts/2022-02-02-decreased-backend-tests-step/index.md

```diff
@@ -0,0 +1,174 @@
+
+## Actions and Results
+
+### Baseline
+
+As a baseline, we considered the total execution time for the backend test step: 22 minutes.
+
+1. Map and split slowest Django apps
+
+We had a bunch of tests spread through several Django apps.
+
+We came up with the following hypothesis: What if we could split the faster/slower apps into different jobs?
+
+We started by measuring the execution time for each Django app. Then, we created manually 2 test jobs: the first would have the slowest tests and the second would have the fastest. We tried to split it equally based on total execution time, ie, the second job with faster tests would have more tests than the first.
+
+We got the following results showing an improvement of about 7 minutes
+
+- Job 1 - 14 min 35 s
+- Job 2 - 14 min 15 s
+
+It showed a valid path which was splitting tests to run in parallel. However, the solution to split them into multiple jobs was not feasible, since we're keeping adding more and more tests daily.
+
+2. No migration before backend tests steps
+
+When we were measuring closer the tests execution time, we realized that the pipeline was taking about 8 minutes to run Django migrations (more than 1400 migration files) since we had a lot of databases and cross-references.
+
+Since we were using [pytest-django](https://pytest-django.readthedocs.io/en/latest/index.html), which takes the current model definition to be used as an in-memory database,
+we could remove the migrations steps.
+
+By removing them, we saw a bunch of tests failing. It turned out that we were using Django migration for data migration too. To fix it, we created several pytest fixtures to provide required data instead of inserting them into the database.
+
+By the end of this step, we could get an improvement of 8 minutes in the total pipeline execution time.
+
+3. Blocking external calls
+
+Our code had a lot of dependencies with external services. During the previous tests investigations, we realized that some of them were not properly mocked the external calls.
+
+We could confirm it by using [pytest-socket](https://github.com/miketheman/pytest-socket). We had 3 different types of external calls:
+
+- External APIs calls (port 80 or 443): 113/15685 test cases
+- ElasticSearch calls (port 9200): 176/15685 test cases
+- Redis cache calls (port 6389): 11132/15685 test cases
+
+
+Before investing time to fix those mocks, we first disabled all those tests to validate whether it was a valid investment for our time.
+It proved worth for External API calls and ElasticSearch. For Redis calls, the impact in the codebase would be huge and the time gained would not be worth it.
+
+After mocking properly them, we could have an improvement of 35% by mocking 1,84% tests properly.
+
+In order to prevent it to happen again, we left [pytest-socket](https://github.com/miketheman/pytest-socket) properly configured in our codebase.
```

Comment:
```suggestion
In order to prevent this from happening again, we configured [pytest-socket](https://github.com/miketheman/pytest-socket) in our codebase.
```